### PR TITLE
Prlib functions + classes

### PR DESCRIPTION
**Pull request by conhlee**
Many prlib functions have been solved. A few class function definitions have been added and class members for PrObjectDatabase and PrModelObject have been fixed / added. The classes SpcFileHeader, PrSceneObject, SpmNode, SpmFileHeader, and SpaFileHeader have been added.

### CODE REVIEW

File: src/prlib/scene.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+#include "database.h"
+
+#include "prlib/prlib.h"
+
+#include "model.h"
+
+#include "common.h"
+
+class SpcFileHeader {
```
Comment (polybiusproxy) [inline src/prlib/scene.h]: Make a new line for the bracket here.

File: include/prlib/prlib.h [hunk 2]
```diff
@@ -69,39 +69,119 @@ void PrCleanupModule();
...
+void PrSetSceneEnv(PR_SCENEHANDLE scene, sceGsDrawEnv1 *drawEnv);
+
 void PrPreprocessSceneModel(PR_SCENEHANDLE scene);
 
-PR_MODELHANDLE PrInitializeModel(PR_MODELHANDLE model, PR_SCENEHANDLE scene);
-PR_ANIMATIONHANDLE PrInitializeAnimation(PR_ANIMATIONHANDLE animation);
-PR_CAMERAHANDLE PrInitializeCamera(PR_CAMERAHANDLE camera);
+PR_MODELHANDLE PrInitializeModel(void* modelImage, PR_SCENEHANDLE scene);
+PR_ANIMATIONHANDLE PrInitializeAnimation(void* animImage);
+PR_CAMERAHANDLE PrInitializeCamera(void* cameraImage);
```
Comment (polybiusproxy) [inline include/prlib/prlib.h]: Why did you remove the typedefs here?

File: src/prlib/prlib.cpp [hunk 3]
```diff
@@ -37,20 +39,41 @@ PR_EXTERN void PrInitializeModule(sceGsZbuf zbuf)
...
 INCLUDE_ASM(const s32, "prlib/prlib", PrCleanupScene);
 
-PR_EXTERN void PrSetSceneFrame(PR_SCENEHANDLE model, sceGsFrame frame)
+PR_EXTERN void PrSetSceneFrame(PR_SCENEHANDLE scene, sceGsFrame frame)
 {
     // Empty
 }
 
-INCLUDE_ASM(const s32, "prlib/prlib", PrSetSceneEnv);
+PR_EXTERN void PrSetSceneEnv(PR_SCENEHANDLE scene, sceGsDrawEnv1 *drawEnv)
```
Comment (polybiusproxy) [inline src/prlib/prlib.cpp]: You can freely use the classes on the .cpp file and get rid of the cast.
Please replace every mention of the typedefs with their respective classes:
```suggestion
PR_EXTERN void PrSetSceneEnv(PrSceneObject *scene, sceGsDrawEnv1 *drawEnv);
```

File: src/prlib/prlib.cpp [hunk 4]
```diff
@@ -68,91 +91,162 @@ INCLUDE_ASM(const s32, "prlib/prlib", PrCleanupCamera);
...
 
-    return &model->m_matrix;
+    return *((PrModelObject*)model)->m_matrix.m;
 }
 
-INCLUDE_ASM(const s32, "prlib/prlib", PrHideModel);
+PR_EXTERN void PrHideModel(PR_MODELHANDLE model)
+{
+    // Clear LSB (hide model flag)
+    ((PrModelObject*)model)->m_posture &= 0xfffffffe;
```
Comment (polybiusproxy) [inline src/prlib/prlib.cpp]:
```suggestion
    model->m_posture &= ~1;
```
This is removing the model show flag.

File: src/prlib/prlib.cpp [hunk 5]
```diff
@@ -68,91 +91,162 @@ INCLUDE_ASM(const s32, "prlib/prlib", PrCleanupCamera);
...
+}
 
 INCLUDE_ASM(const s32, "prlib/prlib", PrSetDefaultCamera);
 
 INCLUDE_ASM(const s32, "prlib/prlib", PrSetAppropriateDefaultCamera);
 
 INCLUDE_ASM(const s32, "prlib/prlib", PrShowModel);
 
-PR_EXTERN NaMATRIX<float, 4, 4>* PrGetModelMatrix(PrModelObject *model)
+PR_EXTERN float* PrGetModelMatrix(PR_MODELHANDLE model)
```
Comment (polybiusproxy) [inline src/prlib/prlib.cpp]: Don't remove NaMATRIX - same on the other functions.

File: src/prlib/prlib.cpp [hunk 6]
```diff
@@ -166,67 +260,155 @@ PR_EXTERN void PrSetStage(int stage)
...
+{
+    if (focalLen != 0.f) {
+        if (focalLen < 0.f || defocusLen <= focalLen) {
+            defocusLen = 0.f;
+            return;
+        }
+    }
+    else {
+        defocusLen = 0.f;
+    }
```
Comment (polybiusproxy) [inline src/prlib/prlib.cpp]: Don't use K&R indentation.

File: src/prlib/prlib.cpp [hunk 7]
```diff
@@ -166,67 +260,155 @@ PR_EXTERN void PrSetStage(int stage)
...
     PrRenderStuff* renderStuff = &prRenderStuff;
     return &renderStuff->m_renderStatistics;
 }
 
-INCLUDE_ASM(const s32, "prlib/prlib", PrSetModelVisibility);
+PR_EXTERN void PrSetModelVisibillity(PR_MODELHANDLE model, u_int nodeIndex, u_int visible)
+{
+    SpmNode* spmNode;
+
+    if (nodeIndex < ((PrModelObject*)model)->m_spmImage->m_nodeNum) {
```
Comment (polybiusproxy) [inline src/prlib/prlib.cpp]: Don't use K&R.

File: src/prlib/prlib.cpp [hunk 8]
```diff
@@ -166,67 +260,155 @@ PR_EXTERN void PrSetStage(int stage)
...
+        spmNode = ((PrModelObject*)model)->m_spmImage->m_nodes[nodeIndex];
+        if (visible) {
+            spmNode->m_flags &= 0xfffdffff;
```
Comment (polybiusproxy) [inline src/prlib/prlib.cpp]:
```suggestion
            spmNode->m_flags &= ~0x20000;
```

File: src/prlib/prlib.cpp [hunk 9]
```diff
@@ -166,67 +260,155 @@ PR_EXTERN void PrSetStage(int stage)
...
+            return;
+        }
+        spmNode->m_flags |= 0x20000;
```
Comment (polybiusproxy) [inline src/prlib/prlib.cpp]:
```suggestion
        if (visible)
            spmNode->m_flags &= ~0x20000;
        else
            spmNode->m_flags |= 0x20000;
```
Can you try if this matches?

**Review by polybiusproxy — CHANGES_REQUESTED** [review]

File: src/prlib/prlib.cpp (hunk 3 above)
Comment (conhlee) [reply src/prlib/prlib.cpp]: This causes a conflict with the header file's definition.

Comment (polybiusproxy) [reply src/prlib/prlib.cpp]: Hmm, perhaps the .cpp file can include another header, while making the functions at `prlib/prlib.h` externs, because I doubt they wrote it like this. It's a statically linked library after all.

File: include/prlib/prlib.h (hunk 2 above)
Comment (conhlee) [reply include/prlib/prlib.h]: There's no suitable type for modelImage (PR_MODELHANDLE is reserved for PrModelObject, not SpmFileHeader) so I did that for PrInitializeModel. Definitely forgot the types for the others though lol

File: src/prlib/prlib.cpp (hunk 5 above)
Comment (conhlee) [reply src/prlib/prlib.cpp]: NaMATRIX and NaVECTOR are C++ structs and are not compatible with the C definitions as far as I know

File: src/prlib/prlib.cpp [hunk 10]
```diff
@@ -262,15 +262,16 @@ PR_EXTERN void PrSetStage(int stage)
...
+        if (focalLen < 0.f || defocusLen <= focalLen)
+        {
             defocusLen = 0.f;
             return;
         }
     }
-    else {
+    else
         defocusLen = 0.f;
-    }
```
Comment (polybiusproxy) [inline src/prlib/prlib.cpp]: Use brackets here too please

File: src/prlib/prlib.cpp [hunk 11]
```diff
@@ -166,33 +260,88 @@ PR_EXTERN void PrSetStage(int stage)
...
+{
+    model->ResetContour();
+}
 
-INCLUDE_ASM(const s32, "prlib/prlib", PrResetPosture);
+PR_EXTERN void PrSavePosture(PrModelObject *model)
+{
+    model->SavePosture();
+}
+PR_EXTERN void PrResetPosture(PrModelObject *model)
```
Comment (polybiusproxy) [inline src/prlib/prlib.cpp]: New line here.

File: src/prlib/prlib.cpp [hunk 12]
```diff
@@ -166,33 +260,88 @@ PR_EXTERN void PrSetStage(int stage)
...
+{
+    return scene->m_fDepthLevel;
+}
 
-INCLUDE_ASM(const s32, "prlib/prlib", PrSavePosture);
+PR_EXTERN void PrSaveContour(PrModelObject *model)
+{
+    model->SaveContour();
+}
+PR_EXTERN void PrResetContour(PrModelObject *model)
```
Comment (polybiusproxy) [inline src/prlib/prlib.cpp]: New line here.

**Review by polybiusproxy — APPROVED** [review]
